src/main/java/com/hrznstudio/titanium/recipe/generator/titanium/ResourceRegistryProvider.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * This file is part of Titanium
+ * Copyright (C) 2019, Horizon Studio <contact@hrznstudio.com>.
+ *
+ * This code is licensed under GNU Lesser General Public License v3.0, the full license text can be found in LICENSE.txt
+ */
+
+package com.hrznstudio.titanium.recipe.generator.titanium;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.hrznstudio.titanium.api.material.IResourceHolder;
+import com.hrznstudio.titanium.block.BlockBase;
+import com.hrznstudio.titanium.item.ItemBase;
+import com.hrznstudio.titanium.material.ResourceRegistry;
+import com.hrznstudio.titanium.recipe.generator.IJSONGenerator;
+import com.hrznstudio.titanium.recipe.generator.IJsonFile;
+import net.minecraft.block.Block;
+import net.minecraft.data.*;
+import net.minecraft.item.Item;
+import net.minecraft.tags.BlockTags;
+import net.minecraft.tags.ItemTags;
+import net.minecraft.util.ResourceLocation;
+
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+public class ResourceRegistryProvider implements IDataProvider {
+
+    private static final Gson GSON = (new GsonBuilder()).setPrettyPrinting().create();
+    private final DataGenerator generator;
+    private ItemTagsProvider itemTagsProvider;
+    private BlockTagsProvider blockTagsProvider;
+
+    public ResourceRegistryProvider(DataGenerator generator) {
+        this.generator = generator;
+        this.itemTagsProvider = new ItemTagsProvider(generator) {
+            @Override
+            protected void registerTags() {
+                ResourceRegistry.getMaterials().forEach(material -> {
+                    material.getGenerated().values().stream().filter(entry -> entry instanceof IResourceHolder).forEach(entry -> {
+                        if (entry instanceof BlockBase) {
```

Comment:
You shouldn't hard code for BlockBase. If you need something, you should turn it into an interface, so that any blocks that are added can be handled here, even if they have to extend something else, like Minecart Tracks or such.